### PR TITLE
Fix parsing of '-' when inside a range

### PR DESCRIPTION
**Pull request by pgovind**
Fixes https://github.com/dotnet/corefx/issues/31786.

This definitely is a breaking change since we're going from parsing patterns such as `[a-\-b]` to throwing an exception (which is the right thing to do).

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
If/when this goes in, can you file issue against roslyn to update their regex parser to be compliant with this change?  Thanks!

**Comment by danmosemsft** [comment]
@CyrusNajmabadi presumably we should be opening Roslyn issues whenever we make a change that causes parsing to fail when it succeeded before or vice versa?

Does your parser respect the regex options (if it can figure them out)? eg https://github.com/dotnet/runtime/pull/993 only changes behavior with the ecmascript option.

**Comment by CyrusNajmabadi** [comment]
> Does your parser respect the regex options (if it can figure them out)? eg #993 only changes behavior with the ecmascript option.

Yup, it does.  

> @CyrusNajmabadi presumably we should be opening Roslyn issues whenever we make a change that causes parsing to fail when it succeeded before or vice versa?

Any time parsing behavior (not runtime behavior) changes at all it would be good to know.  it will affect how we classify things.  So if you would expect classification to change (or parsing to succeed/fail differently), then it would be great to let us know.  Thanks!

**Review by jeffhandley — APPROVED** [review]

**Comment by danmosemsft** [comment]
cc @stephentoub to see this request/comment: https://github.com/dotnet/runtime/pull/1075#issuecomment-568159246

**Comment by stephentoub** [comment]
> cc @stephentoub to see this request/comment

Thanks.  I'm not changing any behaviors around what parses and how.

**Comment by CyrusNajmabadi** [comment]
Well, parsing behavior is changing (i.e. legal code is becoming illegal).  That's 100% fine with me (i opened the bug after all).  We'd just want to follow this up witha  roslyn bug to ensure that regex parser is in sync.  

I'm happy to do all that here. However, for future work, it would be great if you guys could file the bug (in case no one on roslyn is paying attention).

**Comment by stephentoub** [comment]
>>  I'm not changing any behaviors around what parses and how.

> Well, parsing behavior is changing

@CyrusNajmabadi, this isn't my PR :)  Dan cc'd me presumably because I've been submitting other PRs to improve regex perf.

**Comment by danmosemsft** [comment]
> Dan cc'd me presumably because I've been submitting other PRs to improve regex perf, and I was referring to the fact that I'm not changing the parsers behavior in those.

Yup

**Comment by CyrusNajmabadi** [comment]
Gotcha!

**Comment by pgovind** [comment]
I just rebased on upstream/master. I'll merge the PR once CI goes green

File: src/libraries/System.Text.RegularExpressions/tests/RegexParserTests.cs [hunk 1]
```diff
@@ -383,7 +383,7 @@ static RegexParserTests()
         [InlineData(@"[a-\c>]", RegexOptions.None, RegexParseError.UnrecognizedControl)]
         [InlineData(@"[a--]", RegexOptions.None, RegexParseError.ReversedCharRange)]
         [InlineData(@"[--a]", RegexOptions.None, null)]
-        [InlineData(@"[a-\-]", RegexOptions.None, null)]
+        [InlineData(@"[a-\-]", RegexOptions.None, RegexParseError.ReversedCharRange)]
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/tests/RegexParserTests.cs]: Given all of the churn in regex right now, I re-enabled the tests to run against netfx, to help minimize the chances of breaks.  As such, these tests that change expected results will need to be factored out into a separate test that is skipped on netfx.  Sorry for the hassle :)

**Comment by pgovind** [comment]
Thanks @stephentoub !

**Comment by danmosemsft** [comment]
> If/when this goes in, can you file issue against roslyn to update their regex parser to be compliant with this change? Thanks!

@pgovind sounds like you'll want to open one. (Perhaps call out any other parsing changes we made in the last year?)

**Comment by pgovind** [comment]
> sounds like you'll want to open one. (Perhaps call out any other parsing changes we made in the last year?)

Yup yup. I'll open an issue in Roslyn later today

**Comment by CyrusNajmabadi** [comment]
I missed that you were going ot do this @pgovind .  I already opened: https://github.com/dotnet/roslyn/issues/41291

**Pull request merged**